### PR TITLE
CMake build with Visual Studio on Windows

### PR DESCRIPTION
**Pull request by wo80**
This pull request fixes 3 issues when building with Visual Studio on Windows:

1) The `unistd.h` header is not available.
2) `getopt` is not available on Windows:
    This is fixed by adding a CMake module which checks whether `getopt` is installed, so at least examples and tests can be compiled in case it is found.
3) When including `<math.h>`, MSVC by default defines a `complex` type. This is suppressed by adding `_COMPLEX_DEFINED` as a compile flag.

All tests pass (using CMake's `ctest`). Only MSVC should be affected by the changes, but I also checked with GCC.

<del>The pull request should be ready to merge as is, but please review carefully!</del>
Seems to need further discussion ...

### CODE REVIEW

**Comment by gruenich** [comment]
* Do you mind creating a separat merge request for e3185e3? (I think the change is then good to be included)
* Do you mind creating a separat merge request for 011cb2b, 3d58dbc, and d70eaf1 and squash the three commit into one? (I think the change is then good to be included)
* Before exclusion of `unistd.h` (aea9075, f03256d), we should first figure out why we include this on Unix. Maybe we can rely on a common C header?
* Regarding the remainder, the `getops.h` stuff, we should first continue the discussion in #104. It is worth a discussion conditionally excluding tests and examples as an alternative to provide a `getops.h`.

**Comment by wo80** [comment]
> Do you mind creating a separat merge request for https://github.com/xiaoyeli/superlu/commit/e3185e3889548c70d4c8dd74b3be650372ac5ff1?

I don't mind.

> Do you mind creating a separat merge request for https://github.com/xiaoyeli/superlu/commit/011cb2b55f39415da82c8b8982c454957a4f5867, https://github.com/xiaoyeli/superlu/commit/3d58dbcbf8cb6f2039079fc2c158d78d45cef863, and https://github.com/xiaoyeli/superlu/commit/d70eaf1331ed75fd010b59540e9cce77e051c886 and squash the three commit into one?

I don't mind.

> Before exclusion of unistd.h [...], I tried whether we really use this include. Turns out, this is not the case, so I created [...]

I already brought this up in #104 . I think you are right that `unistd.h` isn't needed, but since https://github.com/xiaoyeli/superlu/commit/a49ee5f535c932e6fa604eb12fac3e7796231713 explicitly states, that this was added to make the examples work, I think we should wait and see what @xiaoyeli has to say about it.

Since I was already working on this, I wonder why you needed to create a separate pull request?

> Regarding the remainder, the getops.h stuff [...]

I think the solution I provided using CMake's `find_package` is easy enough. No need to add the getopt code to this project (but still an option, of course).

@xiaoyeli If you think, that separate pull requests should be created, feel free to close this one.

File: cmake/FindGetOpt.cmake [hunk 1]
```diff
@@ -0,0 +1,25 @@
+include(FindPackageHandleStandardArgs)
+
+if (NOT WIN32)
```
Comment (gruenich) [inline cmake/FindGetOpt.cmake:3]: What's the purpose of this path? On Unix the header is present, no need to fiddle around with PKG config - especially as getops won't have a pkg config package on its own.

File: CMakeLists.txt [hunk 2]
```diff
@@ -192,6 +194,10 @@ if(METIS_FOUND)
   set(HAVE_METIS TRUE)
 endif()
 
+if(MSVC)
+  find_package(GetOpt)
```
Comment (gruenich) [inline CMakeLists.txt:202]: Should be called _wingetopt_.

File: cmake/FindGetOpt.cmake [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+find_library(GETOPT_LIBRARY
+    NAMES
+    getopt
+    PATHS
+    ${PKG_LIBGETOPT_LIBRARY_DIRS}
+    /usr/lib
+    /usr/local/lib
+)
+
```
Comment (gruenich) [inline cmake/FindGetOpt.cmake:24]: Should use modern CMake and put everything into a target. Maybe it is worth to create a merge request for wingetopts to export the target. Then SuperLU does not need a FindGetOPs.cmake, but would use wingetops' CMake config file, instead.

**Comment by gruenich** [comment]
> Since I was already working on this, I wonder why you needed to create a separate pull request?

Sorry, I did not want to offend you. I already wrote and tested the code back for # 104 in June, but I forgot to create an MR for this.

I think we should create an MR for the wingetops stuff, and review it there more thoroughly. Maybe we can add stuff to wingetops and not put everything into SuperLU.

File: cmake/FindGetOpt.cmake (hunk 1 above)
Comment (wo80) [reply cmake/FindGetOpt.cmake]: Removed in https://github.com/xiaoyeli/superlu/pull/107/commits/b5dcde6c4a7a9900247492d4a99b753c2fbbc86c

File: CMakeLists.txt (hunk 2 above)
Comment (wo80) [reply CMakeLists.txt]: Renamed in https://github.com/xiaoyeli/superlu/pull/107/commits/b5dcde6c4a7a9900247492d4a99b753c2fbbc86c

File: cmake/FindGetOpt.cmake (hunk 3 above)
Comment (wo80) [reply cmake/FindGetOpt.cmake]: Well, I just wanted to finally get the CMake build work with Visual Studio (issue #104 seemed to be stalled). That's what this PR does. Is there room for optimization? Sure. Is that a reason for further delay. You decide...

**Comment by wo80** [comment]
> Sorry, I did not want to offend you. I already wrote and tested the code back for # 104 in June, but I forgot to create an MR for this.

Thanks for explaining. I was a bit irritated.

I think we agree that this PR won't get merged. Should we still use this for discussion, or should I just create new PR for the getopt stuff?

For me the addition of `_COMPLEX_DEFINED` was necessary to get the code compiled. Where should that change go? Another separate PR?

**Pull request merged**